### ISSUE TITLE
Export fails during the ModelOpt AutoCast stage before TensorRT engine generation begins

### ISSUE THREAD

**Issue by hajgato** [ultralytics/ultralytics#24771]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

**Observed behavior**

Export fails during the ModelOpt AutoCast stage before TensorRT engine generation begins.

The generated dynamic ONNX itself is valid and executes successfully in ONNX Runtime.

**Error**

```
TensorRT: converting ONNX to FP16 mixed precision with ModelOpt AutoCast...

Running ONNX Runtime to obtain reference outputs...

[E:onnxruntime:, sequential_executor.cc:572 ExecuteKernel]
Non-zero status code returned while running Concat node.
Name:'/model.12/Concat'

Status Message:
concat.cc:154 PrepareForCompute
Non concat axis dimensions must match:
Axis 2 has mismatched dimensions of 1 and 2
```

The traceback originates from:

```
ultralytics.utils.export.engine.modelopt_quantize_onnx
→ modelopt.onnx.autocast.convert_to_mixed_precision
→ polygraphy.backend.onnxrt.runner
→ onnxruntime
```

### Environment

**Environment**

* WSL2 Ubuntu
* Python 3.12.3
* Ultralytics 8.4.63
* torch 2.12.0+cu130
* TensorRT 11.0.0.114
* onnx 1.21.0
* NVIDIA RTX 3060 (12 GB)

### Minimal Reproducible Example

**Reproducer**

```python
from ultralytics import YOLO

model = YOLO("yolo26m.pt")

model.export(
    format="engine",
    dynamic=True,
    half=True,
    batch=2,
    imgsz=1280,
)
```

### Additional

**Additional observations**

* `dynamic=False, half=True` succeeds.
* `dynamic=True` ONNX export succeeds.
* The exported dynamic ONNX executes correctly in ONNX Runtime using FP16 inputs:

```python
outputs = sess.run(
    None,
    {"images": np.random.randn(2,3,1280,1280).astype(np.float16)}
)
print([o.shape for o in outputs])
```

Output:

```
[(2, 300, 6)]
```

This suggests the failure occurs specifically during the ModelOpt mixed-precision conversion/validation step.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hajgato, thank you for reporting this to Ultralytics 🚀! This is an automated response to help move things along, and an Ultralytics engineer will also assist you soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many common questions are already covered.

Since this appears to be a 🐛 Bug Report, thank you for already including a reproducible script and environment details 🙌. To help the team investigate the export failure during the ModelOpt AutoCast stage, please also provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible, including:
- The exact install commands used for `ultralytics`, `torch`, `onnx`, `onnxruntime`, TensorRT, and related dependencies
- The full export command output and traceback
- Whether the issue reproduces after upgrading to the latest package version
- Any differences when testing in one of the verified environments below

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24779: Specify calibration data shape for FP16 quantization](https://github.com/ultralytics/ultralytics/pull/24779).

Key changes in this PR that may resolve the problem:
- Extracts the ONNX model’s primary input name earlier from the exported graph
- Reuses that same input name consistently for both INT8 and FP16 quantization flows
- Updates FP16 mixed-precision conversion to pass calibration data into `autocast.convert_to_mixed_precision(...)`
- Provides FP16 conversion with a sample input tensor that matches the model input shape and correct ONNX input name

These updates should make ONNX export quantization in TensorRT/ModelOpt workflows more robust, especially for FP16 mixed-precision conversion.

Credit to @Y-T-G for the PR, with contributions from @onuralpszr.

To test the fix, you can install the latest `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Or, if preferred, wait for the next official PyPI release.

If possible, please try the updated package and let us know whether this resolves the issue on your side.

Thanks 🙏 for reporting this. Further feedback is very welcome if the issue persists.

**Comment by hajgato** [comment]
Yes, the patch https://github.com/ultralytics/ultralytics/pull/24779 solved the problem on my side.

**Comment by onuralpszr** [comment]
@hajgato thank you, we are happy it fixed and works for you. 💯
